cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,995 @@
+          typename F,
+          typename RandomAccessIteartorOut,
+          typename... RandomAccessIteartorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::unrolled_staged>,
+  Offset len,
+  int,
+  F f,
+  RandomAccessIteartorOut out,
+  RandomAccessIteartorIn... ins)
+{
+  constexpr int block_dim        = UnrolledPolicy::BLOCK_THREADS;
+  constexpr int items_per_thread = UnrolledPolicy::ITEMS_PER_THREAD;
+  constexpr int tile_size        = block_dim * items_per_thread;
+  const Offset offset            = static_cast<Offset>(blockIdx.x) * tile_size;
+
+  [&](cuda::std::array<value_t<RandomAccessIteartorIn>, items_per_thread>&&... arrays) {
+  // load items_per_thread elements
+#pragma unroll
+    for (int j = 0; j < items_per_thread; ++j)
+    {
+      const auto idx = offset + (j * block_dim + threadIdx.x);
+      if (idx < len)
+      {
+        // TODO(bgruber): replace by fold over comma in C++17
+        int dummy[] = {(arrays[j] = ins[idx], 0)..., 0}; // extra zero to handle empty packs
+        (void) &dummy[0]; // MSVC needs extra strong unused warning supression
+      }
+    }
+    // process items_per_thread elements
+#pragma unroll
+    for (int j = 0; j < items_per_thread; ++j)
+    {
+      const auto idx = offset + (j * block_dim + threadIdx.x);
+      if (idx < len)
+      {
+        out[idx] = f(arrays[j]...);
+      }
+    }
+  }(cuda::std::array<value_t<RandomAccessIteartorIn>, items_per_thread>{}...);
+}
+
+// used for both, memcpy_async and ublkcp kernels
+template <int BlockThreads>
+struct async_copy_policy_t
+{
+  static constexpr int BLOCK_THREADS = BlockThreads;
+  // items per tile are determined at runtime. these (inclusive) bounds allow overriding that value via a tuning policy
+  static constexpr int MIN_ITEMS_PER_THREAD = 1;
+  static constexpr int MAX_ITEMS_PER_THREAD = 32;
+};
+
+// TODO(bgruber) cheap copy of ::cuda::std::apply, which requires C++17.
+template <class F, class Tuple, std::size_t... Is>
+_CCCL_DEVICE auto poor_apply_impl(F&& f, Tuple&& t, ::cuda::std::index_sequence<Is...>)
+  -> decltype(std::forward<F>(f)(::cuda::std::get<Is>(std::forward<Tuple>(t))...))
+{
+  return std::forward<F>(f)(::cuda::std::get<Is>(std::forward<Tuple>(t))...);
+}
+
+template <class F, class Tuple>
+_CCCL_DEVICE auto poor_apply(F&& f, Tuple&& t)
+  -> decltype(poor_apply_impl(
+    std::forward<F>(f),
+    std::forward<Tuple>(t),
+    ::cuda::std::make_index_sequence<::cuda::std::tuple_size<::cuda::std::__libcpp_remove_reference_t<Tuple>>::value>{}))
+{
+  return poor_apply_impl(
+    std::forward<F>(f),
+    std::forward<Tuple>(t),
+    ::cuda::std::make_index_sequence<::cuda::std::tuple_size<::cuda::std::__libcpp_remove_reference_t<Tuple>>::value>{});
+}
+
+// mult must be a power of 2
+_CCCL_HOST_DEVICE inline auto round_up_to_multiple(int x, int mult) -> int
+{
+  const auto p = cuda::std::popcount(static_cast<unsigned>(mult));
+  assert(p == 1);
```

Comment:
I am not sure. `CUB_DETAIL_DEBUG_ENABLE_DEVICE_ASSERTIONS` is not used anywherre in CUB actually and I want users to get as many error checking as possible when they go for a device debug build.

How is `assert` actually handled in device code? Is `assert` turned on on the device when there is no `NDEBUG` on the host? Or does `assert` work differently on the device?

I want users of an optimized build to not have the assertion, but otherwise they should. I am worried that if I guard the assert behind `CUB_DETAIL_DEBUG_ENABLE_DEVICE_ASSERTIONS`, no user will actually ever turn that on.